HW0/P5/P5.txt

```diff
@@ -1 +1,30 @@
-# Your answers here.
+1. The cashier should divide the bags into pairs and sum each pair: 256 bags / 2 = 128 pairs = 128 sums = 128 seconds.  Then divide those 128 sums into pairs and sum each pair: 64 seconds.  This pattern continues until we have the total.
+
+# of Seconds = 128 + 64 + 32 + 16 + 8 + 4 + 2 + 1 = 255 seconds
+
+2. Yes, they can sum to the total faster.  There are now 8 cashiers.  By dividing the bags into pairs in the same way, 8 sums can be calculated per second.  Thus the first round only takes 128/8 = 16 seconds.  Using the same logic on subsequent rounds:
+
+# of seconds = 16 + 8 + 4 + 2 + 1 + 1 + 1 + 1 = 34 seconds
+
+3.  We still have log-base-2(256) = 8 rounds of summing.  That means the best we can do is 8 seconds (we will always have a worker to do any available sum).
+
+5.  Now, each communicated sum after the first round takes 1 additional second.  
+Since there are still 8 rounds,
+# of seconds = time calculating + time communicating = 8 + 7 = 15 seconds.
+
+6. Now it also takes 1 second to hand a bag to another.
+
+If each cashier takes 1 bag from you:
+# of seconds = time to hand out bags to cashiers + time to communicate first number (before any calculations) + 8 seconds of calculating + 7 seconds communicating sums = 256 + 1 + 8 + 7 = 272 seconds
+
+Note: my answer above assumes that no bags need to be passed after one cashier reads the number on the bag.
+
+If each cashier takes 2 bags from you, only 128 cashiers need to collect bags and no communication occurs before the first calculation.
+Note: this answer assumes that a cashier with 2 bags can instantly read the number from both bags.
+# of seconds = distribution time + calculation time + communication time = 128 + 8 + 7 = 143 seconds
+
+The better strategy though would be for you to hand each cashier that approaches you half of the bags in your possession, while you have more that 2 bags.  When you have 2 bags, give both to the next cashier.  Each cashier with more than 2 bags, should also give half their bags to a bag-less cashier.  
+Here, you would give 128 bags to the first cashier ==> 1 second
+Then you would each give 64 bags to 2 other cashiers ==> 1 second
+Note that there are only 7 rounds of bag distribution since each cashier would keep 2 bags.  Thus, distributing bags would take 7 seconds.
+# of seconds = distribution time + calculation time + communication time =7 + 8 + 7 = 22 seconds
```

Comment:
I am not sure it is correct. The first cashier will take 127 sec to compute the 128 bags which is way more than the 22 seconds that you found. Good explanation though.